Add Sealing APIs (#3652)
openenclave/openenclave: docs/DesignDocs/SealingAPI.md

sealing/unsealing data against an enclave's identity.

## Motivation

Sealing is an important capability of TEEs, which allows an enclave to encrypt
and/or integrity-protect data at rest, using keys (aka., sealing keys) derived
from the enclave's own identity. TEEs may have distinct formulas for key
derivation, and may support different key lengths. And that leads to the desire
for a TEE-agnostic sealing API.

### Objectives

The sealing API should be:

- *TEE-agnostic* - The API should hide TEE specifics to allow reuse of source
  code across TEEs.
- *Accommodative* - Should the developers choose to, the API should allow
  explicit uses of TEE specific features. For example, in the case of SGX, the
  API should allow deriving keys of type `PROVISION_SEAL_KEY`, which may not
  have an equivalent on other TEEs.
- *Easy to use* - It's always a challenge for average developers to encrypt
  data securely by using seal keys directly. The API should shield
  cryptographic complexities from developers, by offering comprehensive
  protection through an intuitive interface.
- *Interoperable with existing SDKs* - The Intel SGX SDK also provides sealing

>   capabilities. It's desirable for enclaves, regardless of the SDKs they built

Reviewer: ```suggestion
  capabilities. It's desirable for enclaves, regardless of the SDKs they are built
```

  with, to be able to exchange sealed blobs, when allowed by the policy.

### Non-Objectives

- Cross-device sealing/unsealing is **not** supported - Sealed blobs created on a device must be unsealed on that same device.
- Cross-TEE sealing/unsealing is **not** supported - Sealed blobs created by an
  enclave in one TEE (e.g., SGX) cannot be unsealed by another enclave in a
  different TEE (e.g., OP-TEE), even if they were both signed by the same
  private key.

## User Experience

At the minimum, two API functions are necessary, namely `oe_seal()` and
`oe_unseal()`. The former encrypts user data into a blob (whose format is
implementation specific) while the latter decrypts/verifies the given blob and

> returns the data back to the user.

Reviewer: ```suggestion
returns the data back to the application.
```

`oe_seal()` requires a seal key to do its job. But how can a seal key be passed to
`oe_seal()`? Here are 3 options:

1. `oe_seal()` takes the seal key directly as a function argument; or
2. `oe_seal()` takes a `key_info` as its argument for deriving the desired seal
   key; or
3. `oe_seal()` takes a `seal_policy` as its argument, and calls
   `oe_get_seal_key_by_policy()` to obtain the seal key.

Option 1 implies access to the seal key by software, which isn't always true
(e.g., TPM never reveals seal keys to software). Option 3 prohibits developers
from fine-tuning `key_info` (probably in a TEE-specific manner), which violates

> the *Accommodative* requirement. Option 2 is our choice.

Reviewer: Which option covers the case where you pass in TEE-agnostic fields (like the seal policy), plus optionally a pointer to a TEE-specific struct?   That's the option that would be my choice.

integrity-protected. But if only integrity needs protection, one could instead
pass `my_data` and `my_data_size` as `additional_data` and
`additional_data_size`, respectively.

```C
uint8_t* blob;
size_t blob_size;
result = oe_seal(key_info,
                 key_info_size,
                 NULL,          // plaintext
                 0,             // plaintext_size
                 my_data,       // additional_data
                 my_data_size,  // additional_data_size
                 &blob,
                 &blob_size);
id (result != OE_OK)
    // Handle encryption error...
else
{
    // Persist blob...

    oe_free(blob); // Free blob when done
}
```

Please note that the blob format is proprietary so there's no architectural way
to extract `additional_data` or `additional_data_size` without unsealing the
blob. That said, the only way to recover `my_data` (and `my_data_size`) is
still via `oe_unseal()`, regardless of whether it was encrypted or not. In other words,
the two code snippets above are identical from a functional perspective but
differ in security. Therefore, developers are strongly encouraged to use
encryption for comprehensive protection.

> `oe_unseal()` decrypts a sealed blob **in-place**. The code snippet below shows

Reviewer: ```suggestion
The code snippet below shows
```
It should not require decryption in place, which would limit applicability.  This also implies you should add free in the sample, just like oe_seal.

> how to decrypt a sealed blob and retrieve a pointer to the plaintext. Please
> note that `my_data` after this call will point to somewhere within `blob`.

Reviewer: ```suggestion
how to decrypt a sealed blob and retrieve a pointer to the plaintext.
```

```C
result = oe_unseal(blob,
                   blob_size,
                   &my_data,        // plaintext
                   &my_data_size,   // plaintext_size
                   NULL,            // additional_data
                   NULL);           // additional_data_size
if (result != OE_OK)
    // Handle decryption error...
else
{
    // Process my_data...
}
```

The initial `key_info` passed to `oe_seal()` could be retrieved by a call to
`oe_get_seal_key_by_policy()`, which takes a `seal_policy` as its only input
parameter. In practice, an enclave may desire to encrypt different things under
different keys. Per
[NIST.SP800-108](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-108.pdf),
*Label* and *Context* should be among the inputs to *KDF* (*Key Derivation
Function*). Most TEEs also support taking addtional *context* or *entropy* into
key derivation. The existing `oe_get_seal_key_by_policy()` API however lacks
that support and prompts for a new API to fill the gap.

The new `oe_get_seal_key_info()` API is introduced with the following

> signature.

Reviewer: We can just pass the fields directly into oe_seal() and not need any get seal key info API.

For `oe_unseal()`, the algorithm was determined by `oe_seal()` and `oe_unseal()` just needs to match it. Neither (1) nor (3) is a concern.

- (1) should not happen simply.
- (3) requires both the sealer and the unsealer to link with the same sealing implementation, which is the ISV's responsibility. A mismatch would fail the unsealer but would **not** compromise security.
- (2) requires support from the sealed blob format and `oe_unseal()`. A simple solution could be to introduce a `revision` field in the blob header when the time comes.
  - [ ] (**Decision?**) Reserve `revision` today? What size?
    - [ ] 8-bit
    - [ ] 16-bit
    - [ ] 32-bit

### How to enable TEE-specific features?

Different TEEs support different features for seal key derivation. E.g., SGX's `KEYREQUEST` contains fields that may not have equivalents on other TEEs, such as `KEYNAME`, `ATTRIBUTEMASK`, `CETMASK`, `MISCMASK`, etc.

(**Decission?**) What to expose at API level?

- [X] Simple API to take "common" parameters and allow fine-tunes in a TEE-specific manner.
  - [X] `entropy` buffer to take in *Label*/*Context* for *KDF*.
  - [ ] `tee_specific` - Not reommended, current proposal.

>   - The rest of `key_info` takes TEE defaults and can be fine-tuned in a TEE specific manner.

Author: @CodeMonkeyLeet, thanks for detailing your idea. I can see that will work, but am not sure what benefits could be brought in.

Each TEE supporting sealing would provide an interface (e.g., an SGX instruction, an OP-TEE API, or a TPM command) that takes in a fix set of parameters. Among them, some are TEE specific (e.g., attributes of an SGX enclave, or a PCR in a TPM) while others are purely crypto related. So there are 2 categories of inputs for deriving keys - TEE-neutral and TEE-specific.

- `oe_seal_setting_agnostic_t` is supposed to abstract TEE-neutral arguments. But what a TEE provides is a KDF essentially, with some of its input supplied by the enclave. What can be supplied by the calling enclave?
  - KDK - The enclave may select one of the KDKs supported by the TEE. This is the `key_policy`.
  - Additional input to KDF that ties to the usage/application of the key, such as a string label, a user name, an algorithm ID, key length, additional entropy, etc. However, they will all be concatenated into a single octet string per NIST.SP800-108 anyway. The size doesn't matter either because it could always be hashed/compressed to the desired size. So we can assert that a single `entropy` (you may argue for a better name) argument will always suffice.
  - With that said, we don't need more TEE-neutral arguments than the 2 above. Then why not pass them directly? Wouldn't that be simpler than encapsulating them in `oe_seal_setting_agnostic_t`?
- `oe_seal_setting_sgx_t` is supposed to abstract SGX specific arguments. But given they are TEE specific, the code cannot be portable. Knowledge leant on one TEE by developers may not be applicable to other TEEs either. Moreover, those developers writing TEE specific code (myself included) are usually familiar with the TEE defined structure (e.g., `KEYREQUEST`). Defining a new abstraction layer basically will force us to write more documents and enclave developers to read more documents. Maintenance will become more costly too, as new setting types will need to be added whenever TEE architecture gets updated. Then why not just use the TEE defined structure instead?

Now back to your first question of why `key_info` is an integral part of `oe_seal()`. I see one responsibility of `oe_seal()` is to translate developers' requests in *logical* form into its *physical* form to be accepted by the underlying TEE. The *logical* form refers to how the key will be used, such as what algorithm (e.g., CMAC vs. GCM vs. HMAC) to be used with the key, what/whose data will be encrypted under they key, etc.; while the *physical* form refers to how the key should be derived, such as how to set up `KEYREQUEST` to be accepted by `EGETKEY`. `key_info` allows most developers to initialize it in a *logical* manner (by `oe_initialize_seal_key_info()`) without preventing others from initializing it in *physical* manner. I consider it the simplest approach to enable TEE-agnostic code without losing any TEE-specific functionality.

>   - The rest of `key_info` takes TEE defaults and can be fine-tuned in a TEE specific manner.

Reviewer: > we don't need more TEE-neutral arguments than the 2 above. Then why not pass them directly? Wouldn't that be simpler than encapsulating them in oe_seal_setting_agnostic_t?

Right, that's what I had suggested too.   However, I don't know if we might find a 3rd agnostic argument in the future and so Simon's proposal of putting the agnostic in a settings struct too is fine with me just in case.

And I am completely with Simon on not having key info which I don't see any value for, it just complicates the developer experience as far as I can tell.

>
> * Actually TEE-agnostic, bakes less implementation directly into OE SDK (or at least, could be more easily broken out as standalone components moving forwards)
> * Accommodative in that the plug-ins get to expose the full range of options specific to a TEE if desired without it filtering through an OE SDK aggregation layer.
>
>   * The trade-off being less expressive sealing semantics for apps that want to span different TEEs, at least in this strawman version.
> * Easy to use in that there's no management of key info at all, and that's incorporated into the `oe_seal` function.
>
>   * With proper built-in defaults per TEE, it could also avoid some of the registration complexity associated with attestation plugins, again because there's a much stronger binding of TEE the enclave is running in vs. attestation.
> * Interoperable with existing Intel SDK, since Intel controls the plugin entirely, they can choose to provide that functionality as they see fit.

> Plug-ins allows multiple implementations to coexist and allows the developer to choose which plug-ins to go into the enclave. Based on the analyses in previous sections, plug-ins are not necessary IMHO.

Reviewer 2: I'll provide a different opinion here which is that the goal of thinking about sealing implementations as plug-ins is more than just allowing multiple implementations to coexist at runtime, it is about defining a consistent set of expectations for how an arbitrary implementation interacts with the OE SDK and other plug-ins. To have something concrete to discuss, consider a formal plugin framework that builds on the alternative public `oe_seal()` proposal above:

```c
/**
 * Extend the types of settings supported by oe_seal to
 * include plug-in specific settings
 */
typedef enum _oe_seal_setting_type
{
    OE_SEAL_SETTING_TYPE_AGNOSTIC = 0xd764778b,
    OE_SEAL_SETTING_TYPE_SGX = 0x8ada353c,
    OE_SEAL_SETTING_TYPE_PLUGIN = 0x887d78bc
} oe_seal_setting_type_t;

typedef struct _oe_seal_setting_plugin
{
    oe_seal_setting_header_t header;
    oe_uuid_t plugin_id;
    const void* custom_data;
    size_t custom_data_size;
} oe_seal_setting_plugin_t;

#define OE_INIT_SEAL_SETTING_PLUGIN(PLUGIN_ID, CUSTOM_DATA, CUSTOM_DATA_SIZE) \
    {                                                                         \
        .header.type = OE_SEAL_SETTING_TYPE_PLUGIN,                           \
        .header.size = sizeof(oe_seal_setting_plugin_t),                      \
        .plugin_id = PLUGIN_ID, .custom_data = CUSTOM_DATA,                   \
        .custom_data_size = CUSTOM_DATA_SIZE                                  \
    }
```

For a basic plug-in model, we would also need some management functions. Note that the rest of this discussion considers these to also be in the public headers, but are primarily of interest to sealing implementation providers:

```c
/**
 * Definition for registering a seal plugin with the OE SDK
 */
typedef struct _oe_seal_plugin_definition
{
    /* UUID for the seal plugin */
    oe_uuid_t plugin_id;

    /* Callback implemented by the plugin to handle oe_seal().
     * Note that seal_callback expands the function prototype of
     * oe_seal() to pass plugin-defined custom_data to the plugin */
    oe_result_t (*seal_callback)(
        const oe_seal_setting_t* settings,
        size_t settings_count,
        const void* custom_data,
        size_t custom_data_size,
        const uint8_t* plaintext,
        size_t plaintext_size,
        const uint8_t* additional_data,
        size_t additional_data_size,
        uint8_t** blob,
        size_t* blob_size);

    /* Callback implemented by the plugin to handle oe_unseal().
     * This has the same function prototype as oe_unseal(). */
    oe_result_t (*unseal_callback)(
        uint8_t* blob,
        size_t blob_size,
        uint8_t** plaintext,
        size_t* plaintext_size,
        uint8_t** additional_data,
        size_t* additional_data_size);
} oe_seal_plugin_definition_t;

/**
 * Register a plugin for oe_seal() and oe_unseal().
 *
 * @param[in] plugin Definition of the plugin to register.
 * @param[in] make_default Sets the plugin to be the default for `oe_seal()`
 * and `oe_unseal()`
 *
 * @retval OE_OK Plugin was successfully registered or already registered.
 * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
 * @retval OE_OUT_OF_MEMORY Failed to allocate memory.
 */
oe_result_t oe_register_seal_plugin(
    oe_seal_plugin_definition_t plugin,
    bool make_default);

/**
 * Unregister a specific sealing plug-in
 *
 * @param[in] plugin_id UUID of the plugin to unregister.
 *
 * @retval OE_OK Plugin was successfully unregistered.
 * @retval OE_NOT_FOUND Specified plugin is not registered.
 */
oe_result_t oe_unregister_seal_plugin(oe_uuid_t plugin_id);
```

As an example, consider a new SGX sealing plugin that wants to allow the developer to specify a label/context value for the plugin to
use as part of its key derivation, instead of reserving that field for its own use as with the existing SGX sealing implementation.
To participate in the plugin framework, it would need provide:

```c
/**
 * Public UUID for the new plugin
 */
#define OE_SEAL_PLUGIN_UUID_CUSTOM_SGX                                    \
    {                                                                     \
        0xd2, 0x5f, 0x43, 0x09, 0x4f, 0x66, 0x40, 0x31, 0xa2, 0xab, 0xf1, \
            0x97, 0x3c, 0xbf, 0x8c, 0xbe                                  \
    }

/**
 * Implementation of its seal_callback
 */ 
oe_result_t sgx_custom_seal(
    const oe_seal_setting_t* settings,
    size_t settings_count,
    const void* custom_data,
    size_t custom_data_size,
    const uint8_t* plaintext,
    size_t plaintext_size,
    const uint8_t* additional_data,
    size_t additional_data_size,
    uint8_t** blob,
    size_t* blob_size)
{
    ...
}

/**
 * Implementation of its unseal_callback
 */ 
oe_result_t sgx_custom_unseal(
    uint8_t* blob,
    size_t blob_size,
    uint8_t** plaintext,
    size_t* plaintext_size,
    uint8_t** additional_data,
    size_t* additional_data_size)
{
    ...
}

/**
 * Public definition of the plugin for the developer to register
 */
extern const oe_seal_plugin_definition_t OE_SEAL_PLUGIN_CUSTOM_SGX = {
    .plugin_id = OE_SEAL_PLUGIN_UUID_CUSTOM_SGX,
    .seal_callback = sgx_custom_seal,
    .unseal_callback = sgx_custom_unseal
};

```

Consider the example use case driving the need to support coexisting sealing implementations, where a developer wants to migrate sealed data from one implementation to another. For this example, let's say it is to migrate data sealed using the default OE SDK implementation to the custom SGX sealing plugin (and for the sake of readability, we will elide the result checking from the function calls):

```c
// Register the plugin definition
oe_result_t result = oe_register_seal_plugin(OE_SEAL_PLUGIN_CUSTOM_SGX, false);

// Unseal the data that was previously sealed by the enclave using the default sealer.
result = oe_unseal(
    ciphertext,
    ciphertext_size,
    &decrypted_text,
    &decrypted_text_size,
    &authenticated_additional_data,
    &authenticated_additional_data_size);

// Define custom configuration as specified by the plugin implementation
char* custom_config =
    "{ \"label\": \"Value to hash into KEYREQUEST.KEYID structure (32-bytes)\" }";
size_t custom_config_size = (size_t)strlen(custom_config) + 1;

// Configure the seal settings array to specify the custom SGX seal plugin
sgx_attributes_t attributes_mask = { 0xFFFFFFFF, 0 };
my_seal_settings_t seal_settings = {
    .sgx_settings = OE_INIT_SEAL_SETTING_SGX(KEYPOLICY_MRENCLAVE, attributes_mask, 0),
    .plugin_settings = OE_INIT_SEAL_SETTING_PLUGIN(
        OE_SEAL_PLUGIN_UUID_CUSTOM_SGX, custom_config, custom_config_size)};

// Seal the data using the custom SGX seal plugin
result = oe_seal(
    (oe_seal_setting_t*)&seal_settings,
    2,
    my_plaintext,
    my_plaintext_size,
    my_additional_data,
    my_additional_data_size,
    &ciphertext,
    &ciphertext_size);

```

It's not an incredible amount of overhead to support this scenario, and it asserts a specific distribution of responsibilities between the components that we can discuss in greater detail:

#### OE SDK requirements

* OE SDK must provide a default sealing implementation for each TEE platform that supports sealing.
  * The default implementation is itself implemented as a plugin and adheres to the same set of expectations as other plugins.
  * The default implementation can be registered via global constructor defined by the OE SDK itself.
  * The default implementation must support all OE-defined seal settings (i.e. `oe_seal_setting_agnostic_t` and `oe_seal_setting_sgx_t` for the default SGX sealer).
* OE SDK tracks a global list of registered sealing plugin definitions for `oe_register_seal_plugin()`/`oe_unregister_seal_plugin()`.
  * As an implementation suggestion, the list may be treated as ordered, with the head definition being the default.
    * Registering with `make_default` effectively inserts the definition at the head instead of the tail of the list.
    * Removing a plugin marked as default will make the next definition default.
    * Removing all plugins or all usable plugins for the current TEE platform will result in `oe_seal()`/`oe_unseal()` returning `OE_NOT_IMPLEMENTED`.
    * Access to the list is guarded by mutex for thread safety.
* `oe_seal()` must scan the list of provided `oe_seal_settings` for any specified plugins.
  * If any plugins are specified, they are invoked in the order in which they are specified and before the default plugin for that platform (if different from the specified plugins).
    * Note that this proposal does not require plugins to be specific to a given TEE; a plugin may be itself implemented to be compatible on different TEEs.
* `oe_unseal()` must invoke all registered plugins beginning with the default, until the `oe_unseal()` operation is handle.
  * This can be optimized a bit by requiring a standard data contract for all sealing plugins to use an OE-defined header that specifies the `plugin_id` of the implementation that produced it.
    * The tradeoff is that OE SDK will not support direct compatibility with sealed blobs produced by other SDKs (e.g. Intel SGX SDK).
    * This is likely acceptable because hardware-sealed blobs are generally not intended to be unsealable on a different device, and may not be unsealable by a different implementation of the enclave depending on the sealing key policy.
      * i.e. the only existing use case is very narrow: migrating data sealed by an Intel SGX SDK enclave using the MRSIGNER policy to an OE SDK enclave _on the same machine_.
    * If a standard sealed blob header is used, OE SDK takes responsibility for parsing the standard header and invoking the appropriate plugin implementation by its UUID. Otherwise, each plugin is invoked in turn and is responsible for its own format probing logic until one plugin accepts the blob.

#### Sealing plugin requirements
* `seal_callback` should be able to function without additional parameters.
  * This includes choosing a good default key policy, ideally consistent with OE SDK's default implementation (e.g. if the default implementation uses `OE_SEAL_POLICY_PRODUCT`, we would encourage plugins to use the closest equivalent default).
* `seal_callback` should use information from `oe_seal_setting_t` where appropriate.
  * It is passed all of the `oe_seal_setting_t` from `oe_seal()` and it must ignore (not error) on any settings it does not support.
    * This is necessary for a developer to write their sealing code once, provide all variants of the custom sealing parameters that may exist on one TEE platform or another.
  * This is preferred over `seal_callback` taking all its parameters from `custom_data` although nothing in the framework prevents this.
  * The plugin has the option to privilege more specific options where they conflict with more general `oe_seal_settings_t`
    * For example, a caller may provide the agnostic `OE_SEAL_POLICY_PRODUCT` setting and also a more specific SGX `KEYPOLICY_CONFIGID | KEYPOLICY_ENCLAVE` setting, and the more specific SGX setting should apply.
* `seal_callback` is solely responsible for defining, versioning and interpreting any `custom_data` it accepts.
  * The parameter is entirely opaque to the OE SDK and is passed directly to the sealing implementation.
  * OE SDK may recommend that the `custom_data` be structured in a standard way (e.g. as JSON in the example), but it provides no enforcement. `custom_data` is entirely a private contract between the caller and the sealing plugin that is not mediated by OE SDK.
* If a plugin is executed in an unsupported environment, it can return `OE_UNSUPPORTED`, and OE SDK will attempt to invoke the next specified plugin (or default for the TEE, if different).
* All plugins must support some form of authenticated additional data for the `additional_data` parameter.
  * This is not a new requirement, just calling out the existing requirement reflected in the original `oe_seal()` function prototype.
    * If we relax this requirement, a sealing plugin should at least be expected to error our with `OE_UNSUPPORTED` instead of silently dropping the `additional_data` parameter (or any standard parameter on the `oe_seal()` API).
  * Note that this proposal places no other constraints on the choice of cryptography used by the plugin. It does not suggest key lengths, key derivation flows, encryption or authenticated hashing algorithms.

Author: @CodeMonkeyLeet, I have the same question as above: What are the benefits that could be brought in by this plugin model? You mentioned that

> it is about defining a consistent set of expectations for how an arbitrary implementation interacts with the OE SDK and other plug-ins. 

But I'm not sure what you meant by "arbitrary implementation". In practice, it's always the sealer that specifies the plugin. And given the unsealer is the same entity (i.e., the same enclave as the sealer, or signed by the same key as the sealer), it inherently knows which plugin to use. I don't see why a plugin needs to interact with any entities other than itself. Wouldn't it be simpler to link to the plugin directly, than to look up the plugin using UUID at runtime?

Reviewer 2: @binxing By arbitrary implementation, I mean that sealing functionality can be implemented and consumed by OE SDK users without inclusion or review by OE SDK itself. Plugin models create a consistent set of guidelines and expectations for the user, 

By definition, the plugin already interacts with the OE SDK, and that doesn't change with the introduction of a formal plugin framework (otherwise, it's just a library and we wouldn't be having a conversation about adding/changing public OE SDK APIs at all). This simply formalizes the relationship between OE SDK and a plugin implementation, so that OE SDK can ensure a consistent user experience for how it interacts with the plugins on the user's behalf (including what it does and does not mediate). 

Regarding whether it's simpler to link the plugin directly, the question is simpler for whom and in what context?
- My priorities in thinking about the public API would be that the semantics for the OE SDK user are clear and easy to use, and I'm willing to trade off having the OE SDK & sealing implementation wrap a bit more work.
- In the case where we require a developer to link separate libraries and then invoke underlying methods provided by different libraries as needed, if you have more than one implementation, then the `oe_seal()` weak alias implementation is determined by link order and that can violate the Principle of Least Astonishment when a change in build definitions create a change in runtime behavior. It is not simpler in that regard.

Author: > By arbitrary implementation, I mean that sealing functionality can be implemented and consumed by OE SDK users without inclusion or review by OE SDK itself. Plugin models create a consistent set of guidelines and expectations for the user,

I agree with you that it's important to have consistent guidelines and expectations. But why does that require plugin models? `oe_seal()` and `oe_unseal()` themselves already set forth guidelines and expectations. After all, the public APIs do nothing except dispatching to the plugin. Then why not have the plugin vendor implement `oe_seal()` and `oe_unseal()` directly?

My another question is: how does that justify the `oe_seal_setting_*_t` structures? As I explained above, there are only 2 TEE neutral arguments - one selects KDK and the other supplies additional input to KDF. The rest are TEE specific and would be better not to be abstracted.

> By definition, the plugin already interacts with the OE SDK, and that doesn't change with the introduction of a formal plugin framework (otherwise, it's just a library and we wouldn't be having a conversation about adding/changing public OE SDK APIs at all). This simply formalizes the relationship between OE SDK and a plugin implementation, so that OE SDK can ensure a consistent user experience for how it interacts with the plugins on the user's behalf (including what it does and does not mediate).

Yes, "consistent user experience" is the key. But that doesn't mean we have to do it at the plugin level. I think "mediate" is the point. Could you provide an example of what to mediate by the top level APIs? If there's nothing to mediate, then there's no need for the top level API.

Plugin models work best in situations where the top level APIs need to translate a single invocation into potentially multiple interactions with a plugin, with a new set of parameters derived from the original ones passed into the top level. It doesn't look to me like the case here. 

Another good thing about plugins is they could be dynamically chosen at runtime. E.g., a TLS implementation could host multiple plugins, each for a different algorithm. Then 2 endpoints of a secure connection can negotiate on a commonly supported algorithm without users' intervention. But it isn't the case for sealing either. 

> Regarding whether it's simpler to link the plugin directly, the question is simpler for whom and in what context?
> - My priorities in thinking about the public API would be that the semantics for the OE SDK user are clear and easy to use, and I'm willing to trade off having the OE SDK & sealing implementation wrap a bit more work.

I'd argue the other way. Users would have to learn how to pass parameters in the form of `oe_seal_setting_*_t` structs. That's not incredibly complicated, but is still more complicated than just looking at API prototypes with self-explanatory arguments. It's more cumbersome to specify TEE specific parameters. Developers would have to study OE docs to figure out what settings correspond to which TEE params. They would have to guess the default values for those params if not specified explicitly. They would have to scratch their heads should they be unable to find an OE setting that corresponds to a param that needs to be set. But if we offer them the TEE interface struct (e.g., `KEYREQUEST` in SGX) directly with an initialization function (i.e. `oe_initialize_seal_key_info()`), all those problems will be gone.

> - In the case where we require a developer to link separate libraries and then invoke underlying methods provided by different libraries as needed, if you have more than one implementation, then the oe_seal() weak alias implementation is determined by link order and that can violate the Principle of Least Astonishment when a change in build definitions create a change in runtime behavior. It is not simpler in that regard.

Just to clarify, developers are not recommended to use `oe_seal()` if multiple implementations are linked. And if they have to do that (e.g., `oe_seal()` was referenced by a library whose source code isn't under the developer's control), the first implementation appearing on the linker command line is always the default. That's a well-defined behavior of `ld` and *weak symbols*. 

The plugin model also faces the same problem, as the developer has to have a means to specify the default plugin. The only difference is how. For weak symbols the order in which libs are supplied determines the default; while in the plugin models the source code (for registering plugins) would probably need changes. I don't see either is better than the other though.

- [ ] (**Decision?**) Plug-in model

### How to specify seal keys?

There are 3 options to specify seal key for `oe_seal()`.

- [ ] `oe_seal()` takes all necessary parameters for deriving `key_info` and the seal key.
  - This gives users a one-stop experience.

>   - But it won't work for (TEE specific) features not exposed as a parameter.

Reviewer 2: This strikes me as creating an incorrect conclusion, since there are clearly more options available for parameterizing `oe_seal()` than indicated by the specific strawman presented. It's not a convincing argument that exposing the `key_info` modified from the existing internal OE data structure is the only way to address this issue. See alternative `oe_seal()` proposal above.

  with, to be able to exchange sealed blobs, when allowed by the policy.

### Non-Objectives

- Cross-TEE sealing/unsealing is **not** supported - Sealed blobs created by an
  enclave in one TEE (e.g., SGX) cannot be unsealed by another enclave in a
  different TEE (e.g., OpTEE), even if they were both signed by the same
  private key.

## User Experience

At the minimum, two API functions are necessary, namely `oe_seal()` and
`oe_unseal()`. The former encrypts user data into a blob (whose format is
implementation specific) while the latter descrypts/verifies the given blob and
returns the data back to the user.

`oe_seal()` requires a seal key to do its job. But how to pass the seal key to
`oe_seal()`? Here are 3 options:

1. `oe_seal()` takes the seal key directly as a function argument; or
2. `oe_seal()` takes a `key_info` as its argument and derives the desired seal
   key; or
3. `oe_seal()` takes a `seal_policy` as its argument, and calls
   `oe_get_seal_key_by_policy()` to obtain the seal key.

Option 1 implies access to the seal key by software, which isn't always true
(e.g., TPM never reveals seal keys to software). Option 3 prohibits developers

> from customing `key_info` (probably in a TEE-specific manner), which violates

Author: E.g., `key_id` on SGX needs to be specified probably for every key derivation.  I'll make it clearer.

    size_t plaintext_size,
    const uint8_t* additional_data,
    size_t additional_data_size,
    uint8_t **blob,
    size_t *blob_size);

oe_result_t oe_unseal(
    uint8_t* blob,
    size_t blob_size,
    uint8_t** plaintext,
    size_t* plaintext_size,
    uint8_t** additional_data,
    size_t* additional_data_size);
```

The code snippet below shows how to encrypt data (i.e., `my_data_size` bytes
starting at `my_data`) using `oe_seal()`.

```C
uint8_t* blob;
size_t blob_size;
result = oe_seal(key_info, key_info_size,
                 my_data,       // plaintext
                 my_data_size,  // plaintext_size
                 NULL,          // additional_data
                 0,             // additional_data_size
                 &blob, &blob_size);
id (result != OE_OK)
    // Handle encryption error...
else
{
    // Persist blob...

    oe_free(blob); // Free blob when done
}
```

Please note that `my_data` above will be both encrypted and
integrity-protected. But if only integrity needs protection, one could instead
pass `my_data` and `my_data_size` as `additional_data` and
`additional_data_size`, respectively.

```C
uint8_t* blob;
size_t blob_size;

> result = oe_seal(key_info, key_info_size,

Author: I have the same tendency, and listed as an open [here](https://github.com/binxing/openenclave/blob/wip/sealing/docs/DesignDocs/SealingAPI.md#shall-aad-additional-authenticated-data-be-supported). Let's vote in our next SIG-Arch meeting.

                 NULL,          // plaintext
                 0,             // plaintext_size
                 my_data,       // additional_data
                 my_data_size,  // additional_data_size
                 &blob, &blob_size);
id (result != OE_OK)
    // Handle encryption error...
else
{
    // Persist blob...

    oe_free(blob); // Free blob when done
}
```

> Please note that the blob format is proprietary so there's no architectural way

Author: *Proprietary* here means the blob format is subject to change, because the cipher and mode may change.  So the users shouldn't assume the internal format of the sealed blob. This I think is a common practice, as TPM spec isn't specific on the cipher/mode and sealed blob format either.

> Please note that the blob format is proprietary so there's no architectural way

Reviewer: then change "proprietary" to "internal"

> Please note that the blob format is proprietary so there's no architectural way

Reviewer: ```suggestion
Please note that the blob format is internal to the OE SDK so there's no architectural way
```

                   &my_data,        // plaintext
                   &my_data_size,   // plaintext_size
                   NULL,            // additional_data
                   NULL             // additional_data_size
                   );
if (result != OE_OK)
    // Handle decryption error...
else
{
    // Process my_data...
}
```

The initial `key_info` passed to `oe_seal()` could be retrieved by a call to
`oe_get_seal_key_by_policy()`, which takes a `seal_policy` as its only input
parameter. In practice, an enclave may desire to encrypt different things under
different keys. Per
[NIST.SP800-108](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-108.pdf),
*Label* and *Context* should be among the inputs to *KDF* (*Key Derivation
Function*). Most TEEs also support taking addtional *context* or *entropy* into
key derivation. The existing `oe_get_seal_key_by_policy()` API however lacks
that support and prompts for a new API to fill the gap.

The new `oe_get_seal_key_info()` API is introduced with the following
signature.

```C
oe_result_t oe_get_seal_key_info(
    oe_seal_policy_t seal_policy,
    uint8_t* entropy,
    size_t entropy_size,

>     uint64_t tee_specific,

Reviewer: ```suggestion
    uint64_t tee_specific_flags,
```

    uint8_t** key_info,
    size_t* key_info_size);
```

If compared to `oe_get_seal_key_by_policy()`, `oe_get_seal_key_info()` differs
in:

> - The addition of `entropy`/`entropy_size` parameters - This allows *Label* or

Author: You sound like talking with `oe_get_seal_key()` taken out from OE APIs.

I'd say, even though they are called *seal keys*, they have a lot more uses than sealing. For example, to integrity protect something, one could pass the sha-256 of that something as `KEYID` and use the resulted `key` as the MAC. One could also "expand" a seal key into some other keys like what `oe_get_private_key_by_policy()` does. And for those reasons, I wouldn't take `oe_get_seal_key()` out. We may consider renaming it to avoid confusions.

  *Context* (or both), or simply more entropy to be mixed into the seal key.

> - The addition of `tee_specific` parameter - This is a TEE specific parameter,

Reviewer: ```suggestion
- The addition of `tee_specific_flags` parameter - This is a TEE specific flags parameter,
```

>   or `OE_SEAL_SGX` to use TEE defaults. See

Reviewer: Above line looks wrong.  OE_SEAL_SGX won't use TEE defaults on non-SGX TEEs.

Author: `OE_SEAL_SGX` won't be defined on non-SGX TEEs and will lead to a compiler error.

  [oe_get_seal_key_info()](#oe_get_seal_key_info()) later in this document for
  more information.
- The removal of `key_buffer`/`key_size` output parameters - Direct use of seal
  keys is error prone so discouraged.

The code snippet below shows how to create a `key_info` suitable for a
provisioning enclave. It adds `PROVISION_KEY` to the attribute mask, and uses
`PROVISION_SEAL_KEY` to exclude CPU SVN from key derivation.

```C
sgx_key_request_t *key_info;
size_t key_info_size;
result = oe_get_seal_key_info(OE_SEAL_POLICY_PRODUCT, NULL, 0,
                              OE_SEAL_SGX | OE_SEALKEY_DEFAULT_FLAGSMASK |
                              SGX_FLAGS_PROVISION_KEY,
                              (uint8_t**)&key_info, &key_info_size);
if (result != OE_OK)
    // Handle errors...
else

>     key_info->key_name = SGX_KEYSELECT_PROVISION_SEAL;

Reviewer: What does this line do?   Why is it necessary to do this on the result rather than passing sufficient inputs to the function in 202?   This looks cumbersome.

Author: `key_name` is a very SGX specific feature and is needed only for provisioning enclaves. 

Generally speaking, each TEE has its distinct features, so it's difficult to define a generic API that works for all TEE-specific features. And it'd be overly complicated even if we could manage to define one. Thus I'm shooting for a "reasonable" API definition that would work for the majority (like 90%+) without blocking the minority. E.g., this corner case of provisioning enclave wouldn't be blocked, even though an extra step would be required.

Reviewer: I'd suggest using the same model as we use for oe_create_enclave().    Allow a TEE-specific buffer to be optionally passed in.

Author: I did consider the TEE-specific buffer model but I decided to do away from that, because it's hard to specify some structure fields and use TEE default for others. E.g., for `KEYREQUEST` on SGX, how to specify a custom `ATTRIBUTEMASK` while using defaults for all other fields? The typical approach, which is also what I'm following here, is to get a structure with all defaults filled in, and modify its fields as needed.

```

## Specification

New files:

- [include/openenclave/bits/seal.h](../../include/openenclave/bits/seal.h)
  contains generic definitions.
- [include/openenclave/bits/sgx/seal.h](../../include/openenclave/bits/sgx/seal.h)
  contains SGX specific definitions.
- [enclave/seal_gcm-aes.c](../../enclave/seal_gcm-aes.c) implements `oe_seal()`
  and `oe_unseal()`.

> - [enclave/crypto/gcm.c](../../enclave/crypto/gcm.c) implements GCM-AES128.

Reviewer: How is crypto-agility supported?

Author: Are you talking about switching to a different cipher implementation (e.g., OpenSSL vs. MbedTLS) or a different cipher algorithm (e.g., AES vs SM4)?

- The SDK picks the cipher algorithm, just like what TPM does.
- @mingweishih is working on switching between MbedTLS vs. OpenSSL. Once that's done, my thought is there'll be 2 gcm.c files, one using MbedTLS and the other OpenSSL. The user would be allowed to specify which crypto lib to link.

Reviewer: +1 to Simon's response.   Key length is the main thing I'm referring to but the same argument can be made about other things.  See https://tools.ietf.org/html/rfc7696 for longer discussion.

Author: Only the weakest link matters in security. Given the processor's 128-bit fuse key, it couldn't add security by using AES-256 as the adversary would simply avoid the 256-bit encryption key but attack the 128-bit fuse key instead.

The idea here is to use the strongest key possible. TEEs may evolve overall time and we'd like to use stronger keys on newer platforms without concerning the users.

Moreover, unsealing can only be done on the same platform by an enclave from the same vendor. There's no reason for a single vendor to use one cipher/mode in one of its enclaves but a different cipher/mode in another. That said, the most of [RFC7696](https://tools.ietf.org/html/rfc7696) don't apply here.

Based on the 2 reasons above, I still don't believe there's a need to support m

Modified files:

- [include/openenclave/enclave.h](../../include/openenclave/enclave.h)
  declares `oe_get_seal_key_info()`, `oe_seal()` and `oe_unseal()`.
- [enclave/core/sgx/keys.c](../../enclave/core/sgx/keys.c) implements
  `oe_get_seal_key_info()`.

### oe_seal()

```C
/**
 * Seal data to an enclave using AEAD (Authenticated Encryption with Additional
 * Data).
 *
 * @param[in] key_info The enclave-specific key information to derive the seal
 * key with.
 * @param[in] key_info_size The size of the \p key_info buffer.
 * @param[in] plaintext Optional buffer to be encrypted under the seal key.
 * @param[in] plaintext_size Size of \p plaintext, must be \c 0 if \p plaintext
 * is \c NULL.
 * @param[in] additional_data Optional additional data to be included in the
 * final MAC.
 * @param[in] additional_data_size Size of \p additional_data, must be \c 0 if
 * \p additional_data is \c NULL.

>  * @param[out] blob On success, receive the pointer to a buffer containing both

Reviewer: ```suggestion
 * @param[out] blob On success, receives the pointer to a buffer containing both
```

 * \p additional_data and encrypted \p plaintext, along with necessary
 * information for unsealing. Freed by oe_free().

>  * @param[out] blob_size On success, receive the size of \p blob.

Reviewer: ```suggestion
 * @param[out] blob_size On success, receives the size of \p blob.
```

 *
 * @retval OE_OK \p plaintext and \p additional_data were sealed to the enclave
 * successfully.
 * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
 * @retval OE_OUT_OF_MEMORY Failed to allocate memory.

>  * @retval OE_CRYPTO_ERROR Error occurred during encryption.

Reviewer: ```suggestion
 * @retval OE_CRYPTO_ERROR An error occurred during encryption.
```

 */
oe_result_t oe_seal(
    const uint8_t* key_info,
    size_t key_info_size,
    const uint8_t* plaintext,
    size_t plaintext_size,
    const uint8_t* additional_data,
    size_t additional_data_size,
    uint8_t **blob,
    size_t *blob_size);
```

`oe_seal()` encrypts and packages user data into a sealed blob using *AEAD*
(*Authenticated Encryption with Additional Data*). A sealed blob is comprised

> of a header, followed immediately by payload, which is the concatenation of

Reviewer: ```suggestion
of a header, followed immediately by a payload, which is the concatenation of
```

cipher text and *AAD* (*Additional Authenticated Data*).

The blob header is TEE specific. On SGX, for interoperability, we adopted the
same format as the Intel SGX SDK. Below gives the definition of

> `oe_sealed_blob_header_t` on SGX.

Reviewer: ```suggestion
`oe_sgx_sealed_blob_header_t` on SGX.
```

Author: `oe_sealed_blob_header_t` isn't SGX specific. It's specific to `oe_seal()` implementation instead. Let me move it to a TEE neutral header file.

Author: I take my words back. `oe_sealed_blob_header_t` is SGX specific because of `key_info`. But if treating `key_info` as just an octet string, then the `oe_seal()` implementation is TEE agnostic.  

How about the following?

In `bits/sgx/seal.h`:

```C
typedef sgx_key_request_t _oe_seal_key_info_t;
```

Btw, what's the naming convention for *internal* structs?

In `bits/seal.h`:

```C
typedef {
    _oe_seal_key_info_t key_info;
    // ...
} oe_sealed_blob_header_t;
```

Given the header is cipher/mode specific, how about renaming it to `oe_sealed_blob_header_gcm_aes`?

Reviewer: If it's SGX specific, then the sgx-specific struct should have sgx in the name.  It can be passed in a TEE-agnostic uint8_t* buffer argument for example.

Author: The bad thing about union is errors can't be caught at compile time. The structure `sgx_key_request_t` won't be defined when the target TEE isn't SGX, so union won't work anyway. What we are trying to do here, using C++ terms, is to instantiate a template, with the argument `_oe_seal_key_info_t` set to `sgx_key_request_t`. Would this `typedef` below be an acceptable solution?

```C
typedef sgx_key_request_t _oe_seal_key_info_t;
```

If not, is C++ allowed in OE SDK? Shall I use C++ template instead?

Reviewer: we need APIs to be usable by C.  (Maybe there's C++ in addition, but you must be able to seal/unseal from C)

```C
typedef struct _sgx_sealed_blob_header
{
    sgx_key_request_t key_info;
    uint32_t ciphertext_size; /* also offset of AAD into payload, as cipher
                                 text is followed immediately by AAD */
    uint8_t reserved[12];     /* must be 0 */
    uint32_t payload_size;    /* total size of cipher text and AAD */
    uint8_t iv[12];  /* must be 0 to be compatible with Intel SGX SDK */
    uint8_t tag[16]; /* Authentication Tag */

> } oe_sealed_blob_header_t;

Reviewer: ```suggestion
} oe_sgx_sealed_blob_header_t;
```
SGX-only types need sgx in the name

> #define ki_entropy key_id

Author: It is to make `oe_seal()`/`oe_unseal()` source code compatible with multiple TEEs. 

Different TEEs use different structure fields for holding user supplied entropy. Here a more generic name `ki_entropy` (`ki` stands for `key_info`) is introduced, so that `oe_seal()` can always use `key_info->ki_entropy` to reference the correct structure member. On SGX, it's `sgx_key_request_t::key_id`.

```

> Even though `oe_sealed_blob_header_t` is TEE specific, the source code of

Reviewer: ```suggestion
Even though the header is TEE specific, the source code of
```

`oe_seal()` (and `oe_unseal()` described later) can still be generic, as long

> as `oe_sealed_blob_header_t` has the following fields.

Reviewer: The header structure above is SGX specific, e.g., the key_info is of length sizeof(sgx_key_request_t) which won't be true on other TEEs, and so iv and tag would be at different offsets on different TEEs, so you can't parse it with TEE agnostic code.  You can create an internal layer that gets the IV for example by calling into a TEE-specific function that parses the header, but you cannot assume that the struct above works on other TEEs.

Author: Of course `key_info` will have different sizes. But at the source level it doesn't matter because the compiler will put in the current offsets. We don't care at the binary level as an enclave can be compiled targeting only one TEE.

- `key_info` - This holds `key_info` to be used by `oe_unseal()` to retrieve
  the same seal key. Its size must match `key_info_size` argument passed in.
- `key_info.ki_entropy` - This holds extra entropy to be mixed into the seal
  key. `ki_entropy` is a macro that expands to `key_id` on SGX, as shown above.
- `iv` - This is the *IV* (*Initial Vector*) used by the underlying cipher.
- `tag` - *AT* (*Authentication Tag*). It must be big enough to accommodate
  *AT*, whose size varies depending on the underlying cipher.
- `payload_size` - Size of the payload that follows the header immediately.
- `ciphertext_size` - Size of the cipher text, or the offset into the payload
   where *AAD* starts (as *AAD* follows the cipher text immediately).

Below shows a possible `oe_seal()` implementation. Please note that GCM is
chosen here. Other cipher modes, such as CCM and OCB, would also work. GCM wins

> the election because

Reviewer: either
a) oe_seal should be cipher mode agile and allow using other cipher modes in the future if needed, or
b) oe_seal should be renamed to be gcm-specific so that other cipher modes could be used in the future if needed without causing confusion.

    OE_CHECK(oe_get_entropy(
        header->key_info.ki_entropy, sizeof(header->key_info.ki_entropy), &k));

    OE_CHECK(oe_get_seal_key(
        (uint8_t*)&header->key_info, key_info_size, &key, &key_size));

    payload = (uint8_t*)(header + 1);
    OE_STATIC_ASSERT(sizeof(header->tag) >= 16);
    OE_CHECK(oe_aes_gcm_encrypt(
        key,
        key_size,
        header->iv,
        sizeof(header->iv),
        additional_data,
        additional_data_size,
        plaintext,
        plaintext_size,
        payload,
        header->tag));

    OE_CHECK(oe_memcpy_s(
        payload + plaintext_size,
        additional_data_size,
        additional_data,
        additional_data_size));

    header->ciphertext_size = (uint32_t)plaintext_size;
    header->payload_size = (uint32_t)(plaintext_size + additional_data_size);

    *blob = (uint8_t*)header;
    *blob_size = size;

done:
    oe_free_seal_key(key, NULL);
    if (result != OE_OK)
        oe_free(header);
    return result;
}
```

The above implementation generates `key_info.ki_entropy` randomly and uses a
constant *IV* of 0<sup>96</sup> on every invocation. The rationale behind that
is that, per

> [NIST.SP800-38D](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf),

Author: Overall, I'm with you that we can do better with a random IV, and listed it as an [open](https://github.com/binxing/openenclave/blob/wip/sealing/docs/DesignDocs/SealingAPI.md#what-cipher-mode-to-use-in-sealing). The main reason for keeping a constant IV is for interoperability with Intel SGX SDK. 

> * Intel doesn't document it's key derivation function for seal key generation using a  NIST certified KDF. That puts a question mark on the FIPS certifiability of the implementation.

I also hope intel could have the KDF certified. However, this isn't something that can be fixed by software.

> * While the `KEYID` is a 32-byte value, it is not representative of the actual entropy with regards to the randomness of the keys. Ultimately, it is just one of many inputs (all of the others remaining constant) and the key space is only 128-bits (16 bytes). How you quantify the probability of a key derivation collision is now something that also needs to be accounted for when considering the NIST requirement.

This will **not** lower the chance of key collision **between enclaves**. But for the same enclave, the same `KEYID` always leads to the same key. Given the same *IV*, it'd be guaranteed to violate NIST requirements. Hence I randomize `KEYID` here.

> * If you take into account the birthday paradox and you're picking keys at random, you don't even get to use all 128-bits of that key space before the probability of a collision exceeds the specified threshold.

Sure. Generally speaking, there shouldn't be more than 2<sup>64</sup> 128-bit keys generated **globally**, or the chance of key collision would be considered non-negligible. The next question is how big an impact it would be should 2 keys collide. A random *IV* would mitigate the problem. And I documented that [here](https://github.com/binxing/openenclave/blob/wip/sealing/docs/DesignDocs/SealingAPI.md#what-cipher-mode-to-use-in-sealing). I personally tend to use a random *IV*, but that'd make OE blobs unable to be unsealed by Intel SGX SDK (Intel blob can still be unsealed by OE). My plan is to have a vote in the next SIG-Arch meeting.

> To be clear, it doesn't seem like a security problem since even if an attacker could get two blobs encrypted with the same key & IV and break AES-GCM, it's not particularly useful if the next blob is encrypted with a different key that you then have to go find another match for. It just makes the compliance question something we need to figure out how to actively defend.

This is a security problem indeed! Same key and IV would reveal to the adversary the GHASH key, which would allow existential forgery.

> Also, even if the protection is largely from randomly changing the key, why not just generate a random IV as well at that point? It's not like you're going to be any worse off from a security perspective.

You vote for random IV. Same do I! let's close it in our next SGX-Arch meeting.

> [NIST.SP800-38D](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf),

Reviewer 2: > This is a security problem indeed! Same key and IV would reveal to the adversary the GHASH key, which would allow existential forgery

Yes, you're absolutely right, I was only thinking about the Disclosure threat vector not the Tampering one, and so the problem is serious.

        key,
        key_size,
        header->iv,
        sizeof(header->iv),
        payload + header->ciphertext_size,
        header->payload_size - header->ciphertext_size,
        payload,
        header->ciphertext_size,
        header->tag));

    if (plaintext)
        *plaintext = payload;
    if (plaintext_size)
        *plaintext_size = header->ciphertext_size;
    if (additional_data)
        *additional_data = payload + header->ciphertext_size;
    if (additional_data_size)
        *additional_data_size = header->payload_size - header->ciphertext_size;

done:
    oe_free_seal_key(key, NULL);
    return result;
}
```

### oe_get_seal_key_info()

```C
/**
 * Get enclave-specific key information that could be passed to
 * oe_get_seal_key_v2() to derive a symmetric encryption key coupled to the
 * enclave platform.
 *
 * @param[in] seal_policy The policy for the identity properties used to derive
 * the seal key.
 * @param[in] entropy Optional parameter containing additional entropy for key
 * derivation. If \c NULL, oe_get_seal_key_info() will generate entropy on
 * behalf of the caller.
 * @param[in] entropy_size Must be \c 0 if \p entropy is \c NULL.

>  * @param[in] tee_specifc \c OE_SEAL_TEE_AGNOSTIC to use TEE defaults. If \c
>  * OE_SEAL_SGX is specified, the rest of \p tee_specifc is interpreted as the

Reviewer: ```suggestion
 * @param[in] tee_specific_flags \c OE_SEAL_TEE_AGNOSTIC to use TEE defaults. If \c
 * OE_SEAL_SGX is specified, the rest of \p tee_specific_flags is interpreted as the
```

 * \a attribute_mask for \c EGETKEY on SGX.
 * @param[out] key_info On success this points to the enclave-specific key
 * information which can be used to retrieve the key by passing it to
 * oe_get_seal_key_v2(). Freed by calling oe_free_seal_key() or oe_free_key().
 * @param[out] key_info_size On success, this is the size of the \p key_info
 * buffer.
 *
 * @retval OE_OK The seal key info was successfully requested.
 * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
 * @retval OE_OUT_OF_MEMORY Failed to allocate memory.
 */
oe_result_t oe_get_seal_key_info(
    oe_seal_policy_t seal_policy,
    uint8_t* entropy,
    size_t entropy_size,

>     uint64_t tee_specifc,

Reviewer: ```suggestion
    uint64_t tee_specific_flags,
```

    uint8_t** key_info,
    size_t* key_info_size);
```

`oe_get_seal_key_info()` duplicates largely the logic of
`oe_get_seal_key_by_policy()`. Please see [User Experience](#user-experience)
earlier in this documentation for an explanation on the intention of this
API.

`seal_policy` selects the type of enclave identity and has the same definition
and supported values as in `oe_get_seal_key_by_policy()`.

`entropy`/`entropy_size` supplies additional data as input to the KDF.
Different TEEs may expect different entropy sizes, e.g., SGX expects 256-bit
entropy. If `entropy_size` isn't the same as expected, `entropy` will be hashed

> before being passed on to the underlying TEE. `entropy` could be `NULL` and if

Reviewer: I don't understand this.  I'd expect it to fail if you give it insufficient entropy.

Author: Each TEE has entropy built in (e.g., SGX has fuse keys). This **extra** entropy is optionally. Per NIST.SP800-108, inputs to a *KDF* may include a *Label* and/or a *Context*. The latter may contain extra entropy, hence the name of this parameter.

that's the case, `entropy_size` must be `0`, and a random bit sequence of
expected size will be used instead.

> `tee_specific` is a 64-bit TEE specific parameter. Its interpretation varies

Reviewer: Either define this as flags, or make it variable length.  If its purpose is to pass in "additional parameters" in general then it should be a byte buffer with a length.
Elsewhere I assumed you meant it to always be flags, but if not then replace my suggestions elsewhere with making it be a buffer and length.  However, line 555 below implies that is not needed, but I would prefer a model where key_info is NOT modifiable, you just pass all the info as input and you get back opaque key info.

from TEE to TEE. It could be an integer on one TEE, or a pointer to some
structure on another. Its purpose is to pass additional parameters that need to
be tuned frequently, e.g., `attribute_mask` on SGX. It doesn't intend to be
comprehensive, as fine tunes can always be done by casting `key_info` into the
TEE specific structure (e.g., `sgx_key_request_t` on SGX) and modifying it
there.

Values of `tee_specific` currently defined are listed in the table below.

|Value|TEE|Definition|
|-----|---|----------|
|`OE_SEAL_TEE_AGNOSTIC`|*All*|No TEE specific features are requested. Signify to use TEE defaults.
|`OE_SEAL_SGX \| <attribute_mask>`|**SGX**|`<attribute_mask>` specifies `sgx_key_request_t::attribute_mask` as input to `EGETKEY`.

`OE_SEAL_TEE_AGNOSTIC` is TEE neutral. Defining it to zero should suffice.

```C
#define OE_SEAL_TEE_AGNOSTIC 0
```

> `OE_SEAL_SGX` should be defined **if and only if** SGX is the target TEE, in

Reviewer: Why do you need this at all then?  The enclave is already compiled as SGX or not, so it seems error prone to have to deal with mismatches between the enclave type and this flag.

Author: What I want is a compiler (rather than a runtime) error if mismatch happens. That said, there will never be mismatches between `tee_specific` flags and enclave type.

order to guarantee a compiler error should the code be compiled targeting a

> different TEE. The value `SGX_FLAGS_INITED` is recommended as it can never be

Reviewer: Is SGX_FLAGS_INITED a literal define from an Intel header file?   (It doesn't start with OE_ hence my question, since if defined in an OE header file it should be OE_SGX_FLAGS_INITIALIZED or similar.)

Author: `SGX_FLAGS_INITED` is an existing macro defined in `bits/sgx/sgxtypes.h`. Shall I submit another pull request to change it?

Reviewer: That would be great if you could, thanks!

{
    int type;      ///< Setting type. See oe_seal_setting_type_t for details.
    uint32_t size; ///< Size of the buffer pointed to by \c value.p
    union {
        uint64_t q; ///< quad-word value. \c size should be set to \c 0.
        uint32_t d; ///< double-word value. \c size should be set to \c 0.
        uint16_t w; ///< word value. \c size should be set to \c 0.
        uint8_t b;  ///< byte value. \c size should be set to \c 0.
        const void*
            p; ///< buffer. \c size should be set to the buffer size in bytes.
    } value;
} oe_seal_setting_t;
```

Where,

- `type` is the setting type. More details follow.
- `size` is the size of the buffer pointed to by `value.p`, otherwise `0`.
- `value` is an union. Which of its members is valid depends on `type`.

Currently, defined TEE neutral settings are summarized in the table blow. See
[include/openenclave/bits/seal.h](../../include/openenclave/bits/seal.h) for
more details.

|Seal Setting                        |Type      |Description
|------------------------------------|----------|-----------
|`OE_SEAL_SETTING_POLICY`            |`uint16_t`|Either `OE_SEAL_POLICY_UNIQUE` (default) or `OE_SEAL_POLICY_PRODUCT`.

> |`OE_SEAL_SETTING_ADDITIONAL_CONTEST`|*Buffer*  |Buffer pointed to by `value.p` of size `size` will be mixed into the seal key. Please note that not every plug-in supports this setting. It's recommended *not* to specify this setting unless absolutely necessary.

Reviewer 2: ```suggestion
|`OE_SEAL_SETTING_ADDITIONAL_CONTEXT`|*Buffer*  |Buffer pointed to by `value.p` of size `size` will be mixed into the seal key. Please note that not every plug-in supports this setting. It's recommended *not* to specify this setting unless absolutely necessary.
```

|`OE_SEAL_SETTING_SGX_XFRMMASK`           |`uint64_t`|Value of `value.q` will be assigned to `KEYREQUEST.XFRMMASK`.
|`OE_SEAL_SETTING_SGX_MISCMASK`           |`uint32_t`|Value of `value.d` will be assigned to `KEYREQUEST.MISCMASK`.
|`OE_SEAL_SETTING_SGX_CONFIGSVN`          |`uint16_t`|Value of `value.w` will be assigned to `KEYREQUEST.CONFIGSVN`.

#### oe_seal()

`oe_seal()` encrypts user-supplied data into an opaque blob, using a key
derived from the calling enclave's identity.

Please see comments inside the code snippet below for details on the function
parameters.

```c
/**
 * Seal data to an enclave using AEAD (Authenticated Encryption with
 * Additioonal Data).
 *
 * @param[in] plugin_id Optional UUID of the plugin to use. If \c NULL, the
 * default plugin will be used.
 * @param[in] settings Optional array of seal settings to be used.
 * @param[in] settings_count The number of settings specified by \p settings.
 * Must be \c 0 if \p settings is \c NULL.
 * @param[in] plaintext Optional buffer to be encrypted under the seal key.
 * @param[in] plaintext_size Size of \p plaintext, must be \c 0 if \p plaintext
 * is \c NULL.

>  * @param[in] additional_data Optional additional data to be MAC'd under the

Reviewer 2: Same as in bits/seal.h ... incomplete sentence?
